allensdk/brain_observatory/filtered_running_speed/__main__.py

```diff
@@ -0,0 +1,954 @@
+    end_index: int,
+        the end index of the stimulus in reference to the full vsync times
+    lowpass: bool (default=True)
+        Whether to apply a 10Hz low-pass filter to the running speed
+        data.
+    zscore_threshold: float
+        The threshold to use for removing outlier running speeds which might
+        be noise and not true signal.
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe with an index of timestamps and the following
+        columns:
+            "velocity": computed running speed
+            "net_rotation": dx in radians
+            "frame_indexes": frame indexes into the raw frame time,
+            "frame_times": the frame times for the timestamp data
+
+    Notes
+    -----
+    Though the angular change is available in the raw data
+    (key="dx"), this method recomputes the angular change from the
+    voltage signal (key="vsig") due to very specific, low-level
+    artifacts in the data caused by the encoder. See method
+    docstrings for more detailed information. The raw data is
+    included in the final output in case the end user wants to apply
+    their own corrections and compute running speed from the raw
+    source.
+    """
+
+    # Identify "wraps" in the voltage signal that need to be unwrapped
+    # This is where the encoder switches from 0V to 5V or vice versa
+    pos_wraps, neg_wraps = _identify_wraps(
+        vsig, min_threshold=1.5, max_threshold=3.5)
+
+    # Unwrap the voltage signal and apply correction for transient spikes
+    unwrapped_vsig = _unwrap_voltage_signal(
+        vsig, pos_wraps, neg_wraps, max_threshold=5.1, max_diff=1.0)
+
+    angular_change_point = _angular_change(unwrapped_vsig, vin)
+    angular_change = np.nancumsum(angular_change_point)
+
+    # Add the nans back in (get turned to 0 in nancumsum)
+    angular_change[np.isnan(angular_change_point)] = np.nan
+    angular_speed = calc_deriv(
+        angular_change,
+        frame_times
+    )  # speed in radians/s
+
+    linear_speed = deg_to_dist(angular_speed)
+
+    # Artifact correction to speed data
+    wrap_corrected_linear_speed = _clip_speed_wraps(
+        linear_speed, frame_times, np.concatenate([pos_wraps, neg_wraps]),
+        t_span=0.25)
+
+    outlier_corrected_linear_speed = _zscore_threshold_1d(
+        wrap_corrected_linear_speed, threshold=zscore_threshold)
+
+    # Final filtering (optional) for smoothing out the speed data
+    if lowpass:
+        polynomial_b, polynomial_a = signal.butter(
+            FILTER_ORDER,
+            Wn=CRITICAL_FREQUENCY,
+            fs=DEFAULT_SAMPLING_FREQUENCY,
+            btype="lowpass"
+        )
+
+        outlier_corrected_linear_speed = signal.filtfilt(
+            polynomial_b,
+            polynomial_a,
+            np.nan_to_num(outlier_corrected_linear_speed)
+        )
+
+    dx_rad = degrees_to_radians(dx_raw)
+
+    frame_indexes = list(range(start_index, end_index))
```

Comment:
Should there be a safety check to make sure that `len(frame_indexes) == len(frame_times)`, in case the user passes inconsistent data in?

Reply:
This has been refactored.